DHT.cpp

```diff
@@ -149,23 +144,42 @@ boolean DHT::read(bool force) {
   {
     // Turn off interrupts temporarily because the next sections are timing critical
     // and we don't want any interruptions.
+    // NOTE: From this point we cannot use function such as delay(), delayMicroseconds(), 
+    //       Those functions among others rely on interrupts.  
     InterruptLock lock;
 
-    // End the start signal by setting data line high for 40 microseconds.
-    digitalWrite(_pin, HIGH);
-    delayMicroseconds(40);
-
+    // End of the start signal.
     // Now start reading the data line to get the value from the DHT sensor.
     pinMode(_pin, INPUT_PULLUP);
-    delayMicroseconds(10);  // Delay a bit to let sensor pull data line low.
+    // from this point we listen, it's sensor's turn to talk.  
+
+
+    // The following step is a busy-wait loop to continue only 
+    // when the data line has been pulled LOW by the sensor.  
+    //
+    // To avoid infinite loops here, we use a counter as a timeout.
+    // 700 iterations is about 4 ms on a 16Mhz processor.  The sensor should
+    // set the pin LOW within 20 to 40 microseconds.
+    // The (F_CPU / 16000000) is a ratio to keep the timeout to approximately
+    // the same period if processor used has different frequency.
+    unsigned long timesup = (F_CPU / 16000000) * 700;  
```

Comment:
Your code is broken for any MCU with `F_CPU < 16000000`. `timesup` will be 0 because, for example, `(15000000 / 16000000) == 0`. I would recommend lowering the denominator to `1000000` (or lower, depending on the minimum supported `F_CPU`) and increasing the multiplier.